tilavarauspalvelu/models/reservation_unit/validators.py

```diff
@@ -0,0 +1,222 @@
+from tilavarauspalvelu.enums import ReservationKind, ReservationTypeChoice, ReservationUnitPublishingState
+from tilavarauspalvelu.models import Reservation
+from utils.date_utils import local_datetime, local_start_of_day
+
+if TYPE_CHECKING:
+    from collections.abc import Collection
+
+    from tilavarauspalvelu.enums import PaymentType
+    from tilavarauspalvelu.models import ReservationUnit, User
+
+__all__ = [
+    "ReservationUnitValidator",
+]
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
+class ReservationUnitValidator:
+    reservation_unit: ReservationUnit
+
+    def validate_reservation_unit_is_published(self) -> None:
+        if self.reservation_unit.publishing_state not in ReservationUnitPublishingState.states_that_are_visible:
+            msg = "Reservation unit is not currently published."
+            raise ValidationError(msg, code=error_codes.RESERVATION_UNIT_NOT_RESERVABLE)
+
+    def validate_reservation_unit_is_reservable_at(self, begin: datetime.datetime) -> None:
+        if not self.reservation_unit.actions.is_reservable_at(moment=begin):
+            msg = "Reservation unit is not reservable at the time of the reservation."
+            raise ValidationError(msg, code=error_codes.RESERVATION_UNIT_NOT_RESERVABLE)
+
+    def validate_user_is_adult_if_required(self, user: User) -> None:
+        if not self.reservation_unit.require_adult_reservee:
+            return
+
+        # AD users are currently never under age since we have blocked students from signing in.
+        if user.actions.is_ad_user:
+            return
+
+        if not user.actions.is_of_age:
+            msg = "Reservation unit can only be booked by an adult reservee"
+            raise ValidationError(msg, code=error_codes.RESERVATION_UNIT_ADULT_RESERVEE_REQUIRED)
+
+    def validate_user_has_not_exceeded_max_reservations(self, user: User, *, ignore_ids: Collection[int] = ()) -> None:
+        if self.reservation_unit.max_reservations_per_user is None:
+            return
+
+        qs = Reservation.objects.filter_for_user_num_active_reservations(self.reservation_unit, user)
+        if ignore_ids:
+            qs = qs.exclude(pk__in=ignore_ids)
+
+        num_active_user_reservations = qs.count()
+        if num_active_user_reservations >= self.reservation_unit.max_reservations_per_user:
+            msg = "Maximum number of active reservations for this reservation unit exceeded."
+            raise ValidationError(msg, code=error_codes.RESERVATION_UNIT_MAX_NUMBER_OF_RESERVATIONS_EXCEEDED)
+
+    def validate_reservation_unit_is_direct_bookable(self) -> None:
+        if self.reservation_unit.reservation_kind == ReservationKind.SEASON:
+            msg = "Reservation unit is not direct bookable."
+            raise ValidationError(msg, code=error_codes.RESERVATION_UNIT_NOT_DIRECT_BOOKABLE)
+
+    def validate_begin_before_end(self, begin: datetime.datetime, end: datetime.datetime) -> None:
```

Comment:
Should this be a _reservation_ validator instead?

Reply:
This is a little tricky to place since it's also used in reservation creation mutation, where a reservation doesn't exist yet, and a reservation validator would need a reservation to validate against. I debated adding a "create validator" but ultimately didn't, since most of these reservation unit validators would then fit there better, but almost always require a "reservation unit" as an input, which would make all of them more verbose.